addon/components/code-base.js

```diff
@@ -1,21 +1,65 @@
 /* global Prism */
 import Component from '@ember/component';
 import { computed } from '@ember/object';
+import { empty } from '@ember/object/computed';
+import { htmlSafe } from '@ember/template';
 
 export default Component.extend({
   classNameBindings: ['languageClass'],
   inline: false,
   language: 'markup',
+  code: null,
+
+  hasBlock: empty('code'),
 
   languageClass: computed('language', function() {
     return `language-${this.get('language')}`;
   }),
 
+  getBlockContent() {
+    return this.blockElement && this.blockElement.textContent;
+  },
+
+  prismCode: computed('code', function () {
+    const code = this.get('hasBlock') ? this.getBlockContent() : this.get('code');
+    const language = this.get('language');
+    if (!code) return '';
+
+    const grammar = Prism.languages[language];
+    if (!grammar) return '';
+
+    const prismCode = Prism.highlight(code, grammar, language);
+    return htmlSafe(prismCode);
+  }),
+
   getElement() {
     return this.element;
   },
 
-  didInsertElement() {
-    Prism.highlightElement(this.getElement());
+  init() {
+    this._super(...arguments);
+    if (typeof document !== 'undefined') {
+      this.blockElement = document.createElement('div');
+    }
+  },
+
+  didRender() {
+    this._super(...arguments);
+    let code = this.get('code');
+    if (this.get('hasBlock')) {
+      code = this.getBlockContent();
+
+      // if block content has changed, force reevaluation of `prismCode`
+      if (code !== this._lastBlockContent) {
```

Comment:
Should  wee maybe use `didUpdateAttrs` here?

Reply:
This is used to detect changes to the component's block content, and not any props passed to the component, so I don't think `didUpdateAttrs` will be called in this case, does it?

Reply:
I'm not sure. I would have to try it out. Just trying to make things as simple as we can

Reply:
I quickly tried with `didReceiveAttrs`, but that definitely does not work. It is called before rendering, so not able to query the DOM correctly. `didRender` should be fine IMO!